pkg/source/kafka/kafka_source.go

```diff
@@ -0,0 +1,361 @@
+		} else if cfg.SASLAlgorithm == "plaintext" {
+			saramaConfig.Net.SASL.Mechanism = sarama.SASLTypePlaintext
+		} else {
+			return nil, fmt.Errorf("invalid SHA algorithm \"%s\": can be either \"sha256\" or \"sha512\"", cfg.SASLAlgorithm)
+		}
+	}
+
+	// validate TLS if required
+	tlsConfig, err := common.CreateTLSConfiguration(cfg.TLSCert, cfg.TLSKey, cfg.TLSCa, cfg.SkipVerifyTLS)
+	if err != nil {
+		return nil, err
+	}
+	if tlsConfig != nil {
+		saramaConfig.Net.TLS.Config = tlsConfig
+		saramaConfig.Net.TLS.Enable = true
+	}
+
+	client, err := sarama.NewConsumerGroup(strings.Split(cfg.Brokers, ","), fmt.Sprintf(`%s-%s`, cfg.ConsumerName, cfg.TopicName), saramaConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to create Kafka client")
+	}
+
+	return newKafkaSourceWithInterfaces(client, &kafkaSource{
+		brokers:          cfg.Brokers,
+		topic:            cfg.TopicName,
+		consumerName:     cfg.ConsumerName,
+		log:              logger,
+		concurrentWrites: cfg.ConcurrentWrites,
+	})
+}
+
+// newKafkaSourceWithInterfaces creates a new source for reading messages from Apache Kafka, allowing the user to provide a mocked client.
+func newKafkaSourceWithInterfaces(client sarama.ConsumerGroup, s *kafkaSource) (*kafkaSource, error) {
+	s.client = client
+	return s, nil
+}
+
+// GetID returns the identifier for this target
+func (ks *kafkaSource) GetID() string {
+	return fmt.Sprintf("brokers:%s:topic:%s", ks.brokers, ks.topic)
+}
+
+func getKafkaVersion(targetVersion string) (sarama.KafkaVersion, error) {
+	preferredVersion := sarama.DefaultVersion
+
+	if targetVersion != "" {
+		parsedVersion, err := sarama.ParseKafkaVersion(targetVersion)
+		if err != nil {
+			return sarama.DefaultVersion, err
+		}
+
+		supportedVersion := false
+		for _, version := range sarama.SupportedVersions {
+			if version == parsedVersion {
+				supportedVersion = true
+				preferredVersion = parsedVersion
+				break
+			}
+		}
+		if !supportedVersion {
+			return sarama.DefaultVersion, fmt.Errorf("unsupported version `%s`. select older, compatible version instead", parsedVersion)
+		}
+	}
+
+	return preferredVersion, nil
+}
+
+// SHA256 hash
+var SHA256 scram.HashGeneratorFcn = func() hash.Hash { return sha256.New() }
+
+// SHA512 hash
+var SHA512 scram.HashGeneratorFcn = func() hash.Hash { return sha512.New() }
+
+type xdgSCRAMClient struct {
+	*scram.Client
+	*scram.ClientConversation
+	scram.HashGeneratorFcn
+}
+
+func (x *xdgSCRAMClient) Begin(userName, password, authzID string) (err error) {
+	x.Client, err = x.HashGeneratorFcn.NewClient(userName, password, authzID)
+	if err != nil {
+		return err
+	}
+	x.ClientConversation = x.Client.NewConversation()
+	return nil
+}
+
+func (x *xdgSCRAMClient) Step(challenge string) (response string, err error) {
+	response, err = x.ClientConversation.Step(challenge)
+	return
+}
+
+func (x *xdgSCRAMClient) Done() bool {
+	return x.ClientConversation.Done()
+}
```

Comment:
All of this is already defined in the target, we can factor it out into `pkg/common/helpers.go`